## Упражнения по библиотеке Numpy

In [1]:
import numpy as np

In [2]:
#создаем случайный массив для заданий 1 и 2,например из целых чисел и размерности 4X5 (можно любой) 
a = np.random.randint(0, 50, size=(4,5))
a

array([[ 8, 30,  8, 22,  2],
       [ 1, 18, 39,  1, 23],
       [28,  6, 48, 19,  9],
       [36, 11, 19,  9,  8]])

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [3]:
a = np.where((a > 3) & (a < 8), -a, a)
# или можем использовать закоментированный код, указанный ниже
# a[(a > 3) & (a < 8)] *= -1
a

array([[ 8, 30,  8, 22,  2],
       [ 1, 18, 39,  1, 23],
       [28, -6, 48, 19,  9],
       [36, 11, 19,  9,  8]])

**2.** Заменить максимальный элемент случайного массива на 0

In [4]:
a[a == np.max(a)] = 0
# или можем использовать закоментированный код, указанный ниже
# a = np.where(a == np.max(a), 0, a)
a

array([[ 8, 30,  8, 22,  2],
       [ 1, 18, 39,  1, 23],
       [28, -6,  0, 19,  9],
       [36, 11, 19,  9,  8]])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [5]:
from itertools import product

# создаем исходные случайные массивы необходимых размернойстей 
A = np.random.randint(0, 5, size=(2, 2))
B = np.random.randint(0, 5, size=(2, 2))
print("Исходный массив A:")
print(A)
print("Исходный массив B:")
print(B)

# Строим прямое произведение массивов 
cartesian_product = list(product(np.reshape(A, A.size), np.reshape(B, B.size)))
print("Gрямое произведение массивов")
cartesian_product

Исходный массив A:
[[4 2]
 [2 2]]
Исходный массив B:
[[2 3]
 [1 3]]
Gрямое произведение массивов


[(4, 2),
 (4, 3),
 (4, 1),
 (4, 3),
 (2, 2),
 (2, 3),
 (2, 1),
 (2, 3),
 (2, 2),
 (2, 3),
 (2, 1),
 (2, 3),
 (2, 2),
 (2, 3),
 (2, 1),
 (2, 3)]

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [6]:
# создаем исходные случайные массивы необходимых размернойстей 10x3 
np.random.seed(7)
A = np.random.randint(0, 5, size=(8, 3))
B = np.random.randint(0, 5, size=(2, 2))
print("Исходный массив A:")
print(A)
print("Исходный массив B:")
print(B)

# создаем матрицу "compared" истинности для сравнения каждого элемента из массива "A" c элементами массива "B" 
compared = (A[..., np.newaxis, np.newaxis] == B)
# формируем массив "result" из строк массива "A", удовлетворяющих условию задачи
result = A[np.all(np.any(compared, axis =(1, 3)), axis = 1)]
print("Результат:")
print(result)

Исходный массив A:
[[4 1 3]
 [3 4 1]
 [0 1 2]
 [2 0 4]
 [0 4 0]
 [3 2 3]
 [4 4 0]
 [0 0 3]]
Исходный массив B:
[[0 2]
 [0 1]]
Результат:
[[0 1 2]
 [2 0 4]
 [0 4 0]
 [4 4 0]
 [0 0 3]]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [7]:
# создаем исходную случайную матрицу 10x3 
a = np.random.randint(0,3, size=(10,3))
print("Исходная матрица:")
print(a)
print("Результат:")
print(a[~np.all(a == a[:,0,None], axis = 1)])

Исходная матрица:
[[2 2 0]
 [2 1 0]
 [0 1 0]
 [0 2 1]
 [0 0 0]
 [2 1 0]
 [1 0 0]
 [1 0 1]
 [1 0 1]
 [2 1 1]]
Результат:
[[2 2 0]
 [2 1 0]
 [0 1 0]
 [0 2 1]
 [2 1 0]
 [1 0 0]
 [1 0 1]
 [1 0 1]
 [2 1 1]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [8]:
# если нужно удалить все повторяющиеся строки
# например, дана следующий "a1" массив 
# создаем исходную случайную матрицу 5x3 
print("Исходный массив")
a = np.random.randint(0,3, size=(5,3))
print(a)

a1, count = np.unique(a, axis = 0, return_counts = True)
print("Массив после полного удаления повторяющихся строк:")
print(a1[count == 1])

# вывести уникальные строки (без их дубликатов)
a2 = np.unique(a, axis = 0)
print("Массив без дубликатов:")
print(a2)

Исходный массив
[[1 1 1]
 [2 2 2]
 [1 2 0]
 [0 1 1]
 [1 0 2]]
Массив после полного удаления повторяющихся строк:
[[0 1 1]
 [1 0 2]
 [1 1 1]
 [1 2 0]
 [2 2 2]]
Массив без дубликатов:
[[0 1 1]
 [1 0 2]
 [1 1 1]
 [1 2 0]
 [2 2 2]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [9]:
import math

# создаем исходную случайную матрицу размером, например, 4x3 
a = np.random.randint(0, 3, size=(4, 3))
# преобразуем исходную матрицу в список
list_a = a.tolist()
print("Исходная матрица:")
print(a)

# реализация без использования NumPy
res1 = math.prod([list_a[i][i] for i in range(min(len(list_a), len(list_a[0]))) if list_a[i][i]!=0])
# или можем использовать закоментированный код, указанный ниже
#from functools import reduce
#res = reduce((lambda x, y: x * y), [list_a[i][i] for i in range(min(len(list_a), len(list_a[0]))) if list_a[i][i]!=0]) 
print("Результат, полученный без помощи NumPy:", res1)

# реализация с помощью NumPy
res2 = np.prod(np.diag(a)[np.diag(a)!=0])
# или можем использовать закоментированный код, указанный ниже
#np.prod(np.diag(a)[np.nonzero(np.diag(a))])
print("Результат, полученный с помощью NumPy:", res2)

Исходная матрица:
[[2 0 0]
 [1 0 0]
 [0 0 2]
 [2 2 1]]
Результат, полученный без помощи NumPy: 4
Результат, полученный с помощью NumPy: 4


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [10]:
# создаем исходные случайные векторы, состоящие из 4 эелементов 
a = np.random.randint(0, 2, size=(4))
b = np.random.randint(0, 2, size=(4))
# преобразуем исходные вектора в списки
list_a = a.tolist()
list_b = b.tolist()
print("Вектор a:", a, "Вектор b:", b)

# реализация без использования NumPy
res1 = (sorted(a) == sorted(b))
# или можем использовать закоментированный код, указанный ниже
#from collections import Counter
#res1 = (Counter(list_a)==Counter(list_b))
print("Результат, полученный без помощи NumPy:", res1)

# реализация с помощью NumPy
res2 = np.all((np.sort(a)==np.sort(b)))
print("Результат, полученный с помощью NumPy:", res2)

Вектор a: [0 1 1 1] Вектор b: [0 0 1 1]
Результат, полученный без помощи NumPy: False
Результат, полученный с помощью NumPy: False


 * __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [11]:
# исходный вектор 
a = np.array([6, 9, 0, 3, 0, 0, 5, 7, 0])
# преобразуем исходный вектор в списки
list_a = a.tolist()
print("Исходный вектор a:", a)

# реализация без использования NumPy
res1 = max([list_a[i+1] for i in range(len(list_a)-1) if list_a[i]==0])
print("Результат, полученный без помощи NumPy:", res1)

# реализация с помощью NumPy
res2 = np.max(a[1:][(a==0)[:-1]])
print("Результат, полученный с помощью NumPy:", res2)

Исходный вектор a: [6 9 0 3 0 0 5 7 0]
Результат, полученный без помощи NumPy: 5
Результат, полученный с помощью NumPy: 5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [12]:
from itertools import groupby

# исходный вектор 
a = np.array([2, 2, 2, 3, 3, 3, 5])
# преобразуем исходный вектор в списки
list_a = a.tolist()
print("Исходный вектор a:", a)

# реализация без использования NumPy. Вариант с циклом
def rle_func(vector):
    elements = []
    num_elem =[] 
    counter = 1
    elem = vector[0]
    for i in range(1, len(vector)):
        if vector[i] == elem:
            counter += 1
        else :
            elements.append(elem)
            num_elem.append(counter)
            elem = vector[i]
            counter = 1
    elements.append(elem)
    num_elem.append(counter)
    return (elements, num_elem)
res1=rle_func(list_a) 
print("Результат, полученный без помощи NumPy #1:", res1) 

#реализация без использования NumPy. Вариант с groupby
res2=([k for k, g in groupby(list_a)], [len(list(g)) for k, g in groupby(list_a)])
print("Результат, полученный без помощи NumPy #1:", res2) 

# реализация с помощью NumPy
ind_of_elem = np.append(np.where(a[1:] != a[:-1]), len(a) - 1)
num_of_elem = np.diff(np.append(-1, ind_of_elem))
res3 = (a[ind_of_elem], num_of_elem)
print("Результат, полученный с помощью NumPy:", res3)

Исходный вектор a: [2 2 2 3 3 3 5]
Результат, полученный без помощи NumPy #1: ([2, 3, 5], [3, 3, 1])
Результат, полученный без помощи NumPy #1: ([2, 3, 5], [3, 3, 1])
Результат, полученный с помощью NumPy: (array([2, 3, 5]), array([3, 3, 1], dtype=int64))


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [13]:
from scipy.spatial.distance import cdist
import time

# создаем исходные случайные векторы
X = np.random.randint(0, 8, size=(3))
Y = np.random.randint(0, 8, size=(3))
# преобразуем исходные векторы в списки
list_X = X.tolist()
list_Y = Y.tolist()
print("Исходная выборка X:", X)
print("Исходная выборка Y:", Y)

#реализация без использования NumPy
start_time_1 = time.time()
res1 =[list(map(lambda a, b: round(((a-list_X[i])**2+(b-list_Y[i])**2)**0.5, 8), list_X, list_Y)) for i in range(len(list_X))]
print("Результат, полученный без помощи NumPy за время:  --- %s секунд ---" % (time.time() - start_time_1))
print(res1)

#реализация с использованием NumPy
start_time_2 = time.time()
dX = X[..., np.newaxis] - X[np.newaxis, ...]
dY = Y[..., np.newaxis] - Y[np.newaxis, ...]
res2 = (np.array([dX, dY])**2).sum(axis=0)**0.5
print("Результат, полученный с помощью NumPy за время:  --- %s секунд ---" % (time.time() - start_time_2))
print(res2)

#реализация с использованием SciPy
XY = np.stack((X, Y), axis=1)
start_time_3 = time.time()
res3 =cdist(XY, XY)
print("Результат, полученный с помощью SciPy за время:  --- %s секунд ---" % (time.time() - start_time_3))
print(res3)

Исходная выборка X: [0 6 6]
Исходная выборка Y: [5 6 5]
Результат, полученный без помощи NumPy за время:  --- 0.0 секунд ---
[[0.0, 6.08276253, 6.0], [6.08276253, 0.0, 1.0], [6.0, 1.0, 0.0]]
Результат, полученный с помощью NumPy за время:  --- 0.0 секунд ---
[[0.         6.08276253 6.        ]
 [6.08276253 0.         1.        ]
 [6.         1.         0.        ]]
Результат, полученный с помощью SciPy за время:  --- 0.0009889602661132812 секунд ---
[[0.         6.08276253 6.        ]
 [6.08276253 0.         1.        ]
 [6.         1.         0.        ]]


В целях более наглядного сравнения быстродействия работы применненных методов по построению матрицы евклидовых расстояний при создании случайных выборок X и Y необходимо увеличить количество элементов в них.
При формировании матриц евклидовых расстояний наилучший результат по скорости работы показывает пакет SciPy, второе место по данному параметру занимает NumPy. На третьем же месте расположено выполение задачи с помощью базовых функций Python.
   
Резюмируем:
1) SciPy,
2) NumPy,
3) Python.

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [14]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [15]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [16]:
average_calories = calorie_stats.mean().round(2) # среднее количество калорий у конкурентов
average_calories_diff = (average_calories - 60).round(2) # разница между средним количеством калорий у конкурентов и количеством калорий в порции CrunchieMunchies
print(f"Средний калораж: {average_calories}. Разница между средним калоражем и CrunchiesMunchies: {average_calories_diff}")

Средний калораж: 106.88. Разница между средним калоражем и CrunchiesMunchies: 46.88


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [17]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [18]:
median_calories = np.median(calorie_stats_sorted)
median_calories

110.0

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [19]:
all_percentiles = np.percentile(calorie_stats_sorted, [i for i in range(1, 101)])
percentiles_greater_60 = all_percentiles[all_percentiles > 60]
percentiles_greater_60

array([ 70. ,  70. ,  75.6,  83.2,  90. ,  90. ,  90. ,  90. ,  90. ,
        90. ,  90. ,  90. ,  91.6,  99.2, 100. , 100. , 100. , 100. ,
       100. , 100. , 100. , 100. , 100. , 100. , 100. , 100. , 100. ,
       100. , 100. , 100. , 100. , 100. , 100. , 100. , 100. , 106.4,
       110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. ,
       110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. ,
       110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. ,
       110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. , 110. ,
       110. , 115.2, 120. , 120. , 120. , 120. , 120. , 120. , 120. ,
       120. , 120. , 120. , 120. , 120. , 124. , 130. , 130. , 136.8,
       140. , 140. , 140. , 147.2, 150. , 152.4, 160. ])

In [20]:
# значение 70 присутствует в изначальной выборке
percentile_rank = 101 - percentiles_greater_60.size
percentile_rank

4

In [21]:
# искомый процентиль
fourth_percentile = percentiles_greater_60[0]
fourth_percentile

70.0

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [22]:
percentage_greater_60 = (np.mean(calorie_stats_sorted > 60) * 100).round(2)
percentage_greater_60

96.1

In [23]:
a = np.array([1,0,1,1,1,0,1,1,0,1])
a.mean()

0.7

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [24]:
calorie_std = calorie_stats_sorted.std().round(2)
calorie_std

19.36

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Полученные данные в целом дают представление о разбросе калоража у злаков конкурентов CrunchieMunchies и то, как в эту выборку вписывается калорийность самих CrunchieMunchies.
Однако часть полученных метрик, таких как медианное значение 110 калорий (которое показывает, что половина участвующих в анализе злаков имеет калорийность более 110), значение 4-го процентиля (которое отражает, что калораж CrunchieMunchies попадает в 4-х процентную выборку наименьших по значению калорийности злаков), а также стандартное отклонение 19,36 кал. (как мера рассеивания случайной величины относительно среднего значения) - не будут понятны большинству покупателей злаков, поэтому их довольно сложно использовать для маркетинговых целей. Однако эти данные хорошо иллюстрируют исследователю, то преимущество по количеству калорий, которое имеют CrunchieMunchies по сравнению с конкурентами.
Для маркетинговых целей удобнее всего будет использовать процент хлопьев, в которых содержится более 60 калорий на порцию, который составляет 96,1% (хотя значение 4-го процентиля также помогает понять схожую информацию). Вдобавок, можно будет использовать разницу в 46,88 калорий между средним количеством калорий у конкурентов  и калоражем злаков CrunchieMunchies. Показатель не отражает реальную картину разброса данных, однако на покупателя может возыметь должный эффект. В ходе маркетинговой компании упоминание данных величин, например в любой рекламе, быстрее всего бросится покупателю на глаза, а их высокое значение может помочь в привлечении необходимой клиентуры.
В дополнение нужно отметить, что для маркетинга следует подумать и о других преимуществах продукта, о том, что выгодно отличает его от конкурентов, ведь остальные, приблизительно 4%, конкуренты не уступают CrunchieMunchies по исследованному показателю. Калорийность продукта части их них даже меньше, чем у CrunchieMunchies.